packages/langium/src/utils/promise-util.ts

```diff
@@ -73,7 +73,46 @@ export async function interruptAndCheck(token: CancellationToken): Promise<void>
         lastTick = current;
         await delayNextTick();
     }
-    if (token. isCancellationRequested) {
+    if (token.isCancellationRequested) {
         throw OperationCancelled;
     }
 }
+
+/**
+ * Utility class to execute mutually exclusive actions.
+ */
+export class MutexLock {
+
+    private previousAction = Promise.resolve();
+
+    /**
+     * Performs a single async action, like initializing the workspace or processing document changes.
+     * Only one action will be executed at a time.
+     */
+    async lock(action: () => Promise<void>): Promise<void> {
+        // Append the new action to the previous action. We usually don't have to wait for long, as the previous action
+        // 1. has either completed
+        // 2. has been cancelled due to the new write request
+        return this.previousAction = this.previousAction.then(() =>
+            action().catch(err => {
+                if (!isOperationCancelled(err)) {
+                    console.error('Error: ', err);
+                }
+            })
+        );
```

Comment:
Suggestion:
```typescript
        return this.previousAction = this.previousAction.then(
            action,
            err => {
                if (!isOperationCancelled(err)) {
                    console.error('Error: ', err);
                }
            }
        );
```

Reply:
This actually breaks the error recovery, since this isn't the same as `action.catch(err)`